code/cfile/cfilecompression.cpp

```diff
@@ -0,0 +1,219 @@
+	case SEEK_CUR: goal_position = cf->raw_position + offset; break;
+	case SEEK_END: goal_position = cf->size + offset; break;
+	default: return 1;
+	}
+
+	// Make sure we don't seek beyond the end of the file
+	CAP(goal_position,(size_t)0, cf->size);
+
+	cf->raw_position = goal_position;
+
+	return 0;
+}
+
+//Special fseek for compressed files handled by FSO, only SEEK_SET and SEEK_END is supported.
+int fso_fseek(CFILE* cfile, int offset, int where)
+{
+	size_t goal_position;
+
+	switch (where) {
+	case SEEK_SET: goal_position = offset + cfile->lib_offset; break;
+	case SEEK_END: goal_position = cfile->compression_info.compressed_size + offset + cfile->lib_offset; break;
+	default: return 0;
+	}
+	
+	// Make sure we don't seek beyond the end of the file
+	CAP(goal_position, cfile->lib_offset, cfile->lib_offset + cfile->compression_info.compressed_size);
+	
+	int result = 0;
+	if (cfile->fp)
+	{
+		result = fseek(cfile->fp, (long)goal_position, SEEK_SET);
+		Assertion(goal_position >= cfile->lib_offset, "Invalid offset values detected while seeking! Goal was " SIZE_T_ARG ", lib_offset is " SIZE_T_ARG ".", goal_position, cfile->lib_offset);
+	}
+	return result;
+}
+
+void lz41_create_ci(CFILE* cf, int header)
+{
+	cf->compression_info.header = header;
+	cf->compression_info.compressed_size = cf->size;
+	fso_fseek(cf, -12, SEEK_END);
+	auto fNumoffsets = fread(&cf->compression_info.numOffsets, 4, 1, cf->fp);
+	auto fSize = fread(&cf->size, 4, 1, cf->fp);
+	auto fBsize = fread(&cf->compression_info.block_size,4, 1, cf->fp);
+
+	#if !defined(NDEBUG)
+	if (cf->compression_info.numOffsets <= 0 || cf->size <= 0 || cf->compression_info.block_size <= 0 || fNumoffsets < 1 || fSize < 1 || fBsize < 1)
+	{
+		//mprintf(("\n(CI) File: %s . Something went wrong reading the setup data, file is possibly in the wrong format.\n", cf->original_filename.c_str(), cf->compression_info.numOffsets, cf->compression_info.block_size));
+		Assertion(cf->compression_info.numOffsets > 0, "Invalid number of offsets, compressed file is possibly in the wrong format.");
+		Assertion(cf->size > 4, "Invalid filesize, compressed file is possibly in the wrong format.");
+		Assertion(cf->compression_info.block_size > 16, "Invalid block size, compressed file is possibly in the wrong format.");
+	}
+	#endif
+	cf->compression_info.decoderBuffer = (char*)malloc(LZ4_compressBound(cf->compression_info.block_size));
+	cf->compression_info.lastDecBlockPos = 0;
+	cf->compression_info.lastDecBytes = 0;
+	lz41_load_offsets(cf);
+}
+
+void lz41_load_offsets(CFILE* cf)
+{
+	int maxBlocks = (int)cf->size + 8 / cf->compression_info.block_size;
+	cf->compression_info.offsets = (int*)malloc(maxBlocks);
+	int block;
+	int* offsetsPtr = cf->compression_info.offsets;
+
+	fso_fseek(cf, (-4 * (cf->compression_info.numOffsets + 1)) - (sizeof(int)*2), SEEK_END);
+	for (block = 0; block <= cf->compression_info.numOffsets; ++block)
+	{
+		auto bytes_read = fread(offsetsPtr++, sizeof(int), 1, cf->fp);
```

Comment:
You should probably check `cf->compression_info.numOffsets * sizeof(int) <= maxBlocks` with an `Assertion()` to prevent out-of-bounds writes here.